### ISSUE TITLE
How to enable duplex communication between RPI and node?

### ISSUE THREAD

**Issue by tarunsmalviya** [openthread/ot-br-posix#315]
I have built border router using RPi as per guidance provided on website. 
Now, SED is able to send data to CoAP server running on RPi, but CoAP server is not able to send data to SED. 

Can anyone help me achieving it?
Thanks in advance.

**Comment by jwhui** [comment]
@tarunsmalviya, can you provide more details on your setup?

Also, can you provide a packet capture from the wpan0 interface that can be viewed in Wireshark?

**Comment by tarunsmalviya** [comment]
@jwhui I have configured RPi as border router as explained on openthread.io website. I have created a dummy interface having IP 10.0.0.1. A SED is sending data to this address and a CoAP server(californium) is listening to this on 5683 port.

Kindly let me know what extra information you want to know.

PFA for wpan0 and nat64 interface packet capture.
[wpan_trace_2019_07_15_11_20_06.pcap.gz](https://github.com/openthread/ot-br-posix/files/3414607/wpan_trace_2019_07_15_11_20_06.pcap.gz)
[nat_trace_2019_07_15_11_20_06.pcap.gz](https://github.com/openthread/ot-br-posix/files/3414608/nat_trace_2019_07_15_11_20_06.pcap.gz)

**Comment by jwhui** [comment]
@tarunsmalviya, from the packet traces, I see CoAP responses getting sent to Thread devices (i.e. destination with `fddf:dd::/64` prefix). Are you saying that these CoAP responses are not reaching the Thread devices?

**Comment by tarunsmalviya** [comment]
@jwhui No, the responses are reaching Thread devices. 
My concern is that when the Thread node send packet to the RPi, then RPi in return send response and the response reaches the Thread node. But when I try to send packet from RPi to Thread node using IPv6 address of Thread node, the packet is not received by Thread node.

**Comment by jwhui** [comment]
@tarunsmalviya, do the packet traces above have examples of CoAP Request messages being sent from the RPi to Thread devices? If so, can you call out which packet numbers? If not, can you provide a packet trace that shows the failure case?

**Comment by tarunsmalviya** [comment]
@jwhui I will do the same and update you asap.
Also, I have created a dummy loopback network interface having IP 10.0.0.1. CoAP server is listening on  this IP 10.0.0.1 at 5683 port. So, Thread nodes send packet to this dummy interface by appending prefix to 10.0.0.1.

Is it possible to assign a fix IPv4 and IPv6 to wpan0 interface, so that I can bind CoAP server to this IP.

**Comment by jwhui** [comment]
@tarunsmalviya, I am not familiar with Californium implementation details. Does Californium support sending CoAP requests to IPv6 multicast destination addresses?

**Comment by tarunsmalviya** [comment]
@jwhui Let me check with Californium community.

**Comment by jwhui** [comment]
@tarunsmalviya, were you able to resolve this issue?

**Comment by tarunsmalviya** [comment]
@jwhui I am working on it. 
For reference: https://github.com/eclipse/californium/issues/1076#issuecomment-539601929

**Comment by PanzerHabba** [comment]
Hi.
I am also not able to send multicast messages from rpi to thread network.
As @tarunsmalviya , I`m able to send unicast messages.

I am using [node-coap](https://github.com/mcollina/node-coap).
Here is my coap-script running on rpi: (masked ipv6 address of wpan0)

`const coap = require("coap");
var dgram = require("dgram");

const SERVER_PORT = 5683; // Port for UDP connection
const SERVER_HOST = "fdde:ad00:beef:0:ca64:xxx:xxx:3680"; // IPv6 address of wpan0-interface
const IP_V6_SOCKET = "udp6"; // Specify ipv6 connection
const MULTICAST_LINK_LOCAL = "ff02::1";
const MULTICAST_REALM_LOCAL = "ff03::1";

const serverOptions = {
  type: IP_V6_SOCKET
};

const sock = dgram.createSocket("udp6");
const coap_server = coap.createServer(serverOptions);

/* callback for coap_message_received on wpan0-interface */
coap_server.on("request", (req, res) => {
  console.log("Data: ", req.payload);
  res.code = "2.00";
  res.end();
});

coap_server.on("error", err => {
  console.log(`server error:\n${err.stack}`);
  coap_server.close();
});

sock.bind(SERVER_PORT, SERVER_HOST);
coap_server.listen(sock, () => {
  console.log("COAP Server listening on " + SERVER_HOST + ":" + SERVER_PORT);
  incoming_message_count = 0;

  const options = {
    host: MULTICAST_REALM_LOCAL,
    multicast: true,
    multicastTimeout: 2000,
    port: 5683,
    method: "GET",
    pathname: "/test-resource"
  };
  var req = coap.request(options);

  req.on("end", () => {
    console.log("request sent");
    process.exit(0);
  });

  req.end();
});
`

**Comment by PanzerHabba** [comment]
I can also add that multicast-packets sent from node i picked up by tcpdump on rpi.
But multicast packet going out from wpan0 is not detected (e.g. not sent)..
![image](https://user-images.githubusercontent.com/17178340/68845273-776d4900-06cb-11ea-8423-3eb893a3ba15.png)

**Comment by PanzerHabba** [comment]
I finally got this working.
I had to specifically set multicast-interface on node.js dgram-socket.
link: [setMulticastInterface](https://nodejs.org/api/dgram.html#dgram_socket_setmulticastinterface_multicastinterface)

So, my udp-socket will be like this:
![image](https://user-images.githubusercontent.com/17178340/68866281-74d31980-06f4-11ea-9d50-cba2d6d55224.png)


I was assuming this to be automatically set up, when calling bind on wpan0-ipv6/port.
Then I`ll guess that my issue ha nothing to do with openthread.
